### ISSUE TITLE
`ReferenceError: window is not defined` in 0.32.0

### ISSUE THREAD

**Issue by CharlesStover** [ezolenko/rollup-plugin-typescript2#339]
#### Environment

* Yarn 3

##### Versions

The command to use `envinfo` throws an error, so I'm resorting to copy-paste. All versions are latest.

```text
    "rollup": "^2.75.5",
    "rollup-plugin-typescript2": "^0.32.0",
    "typescript": "^4.7.2"
```

The error resides here:

```js
if (typeof commonjsRequire === "function") {
  try {
    lodash = {
      clone: require("lodash/clone"),
      constant: require("lodash/constant"),
      each: require("lodash/each"),
      filter: require("lodash/filter"),
      has:  require("lodash/has"),
      isArray: require("lodash/isArray"),
      isEmpty: require("lodash/isEmpty"),
      isFunction: require("lodash/isFunction"),
      isUndefined: require("lodash/isUndefined"),
      keys: require("lodash/keys"),
      map: require("lodash/map"),
      reduce: require("lodash/reduce"),
      size: require("lodash/size"),
      transform: require("lodash/transform"),
      union: require("lodash/union"),
      values: require("lodash/values")
    };
  } catch (e) {
    // continue regardless of error
  }
}

if (!lodash) {
  lodash = window._;
}
```

The line saying to use `window._` throws an error because `window` does not exist in the Node environment.

This is newly introduced in 0.32.0. Reverting to 0.31.0 fixes this.

**Comment by agilgur5** [comment]
> The error resides here:

This code exists in `0.31.0` as well, can see [the `dist` diff](https://github.com/ezolenko/rollup-plugin-typescript2/commit/4d5debaa835be2a6faa2ad444c8a2af52aec1724#diff-2850176f5f74159335da7582dec86c4c0a84030bc88a957627f80e42d9e16000) (it's in both `+` and `-` as the code merely moved around).
The code is also from `lodash` internals, and not rpt2 source code.

> The line saying to use `window._` throws an error because `window` does not exist in the Node environment.

rpt2 builds itself with itself, which is done with Rollup and Node, so I am unable to reproduce this in a Node environment.

> This is newly introduced in 0.32.0. Reverting to 0.31.0 fixes this.

The only `lodash` related PR in `0.32.0` is #328 , which basically just _reduces_ the usage of `lodash`. It's still used in a few places though and still part of the bundle as such. It's the same version though and it is [pinned](https://github.com/ezolenko/rollup-plugin-typescript2/blob/44116044750b01c3b0d612e36f453cc8e64f79b4/package.json#L61) to a specific version too (not that `lodash` gets many updates nowadays).

I can't really pinpoint what would cause any change here, and you didn't provide a reproduction or a stacktrace / `verbosity: 3` log per the issue template, so can't really investigate further either.

**Comment by CharlesStover** [comment]
I unfortunately didn't snapshot it at a point in time, because I'm trying to do a larger effort around dependency upgrades. This is in a private repository anyway. :(

Here's the stack trace:

```
PS ...\REPO> yarn rollup   
[!] ReferenceError: window is not defined
ReferenceError: window is not defined
    at Object.<anonymous> (...\REPO\.yarn\__virtual__\rollup-plugin-typescript2-virtual-387384c206\0\cache\rollup-plugin-typescript2-npm-0.32.0-9a7f8c5b22-481b4288e8.zip\node_modules\rollup-plugin-typescript2\dist\rollup-plugin-typescript2.cjs.js:20377:3)
    at Module._compile (node:internal/modules/cjs/loader:1103:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
    at require$$0.Module._extensions..js (...\REPO\.pnp.cjs:21415:33)
    at Object.require.extensions.<computed> [as .js] (...\REPO\.yarn\cache\rollup-npm-2.75.5-7f19e4b2e6-fa3e61959e.zip\node_modules\rollup\dist\shared\loadConfigFile.js:617:17)
    at Module.load (node:internal/modules/cjs/loader:981:32)
    at Function.require$$0.Module._load (...\REPO\.pnp.cjs:21255:14)
    at Module.require (node:internal/modules/cjs/loader:1005:19)
    at require (node:internal/modules/cjs/helpers:102:18)
    at Object.<anonymous> (...\REPO\.yarn\__virtual__\@monorepo-template-rollup-config-virtual-198f84a7ca\0\cache\@monorepo-template-rollup-config-npm-2.1.0-f4bb5fa311-4072763686.zip\node_modules\@monorepo-template\rollup-config\src\utils\rollup-config.cjs:1:21)
```

The `@monorepo-template/rollup-config` dependency that calls `typescript2` can be found [here](https://github.com/monorepo-template/dependencies/blob/main/packages/rollup-config/src/utils/rollup-config.cjs#L143).

When I finish the upgrade overall, I'll take `rollup-plugin-typescript2` off `~0.31.0` and put it back on `^0.32.0` and see if it still occurs.

It may be related to the workspace package having `"type": "module"` in its `package.json`, as I notice the dependency is `.cjs`. I didn't experience this in another repo of mine that is `"type": "commonjs"`.

**Comment by agilgur5** [comment]
> Here's the stack trace:

Thanks for providing this. It looks like you're using Yarn PnP too, which is possibly related too?
We also normally ask to run with `clean: true` as well, but in this case, I think that wouldn't make a difference as this happens during parsing of rpt2's `dist` code.

> It may be related to the workspace package having `"type": "module"` in its `package.json`, as I notice the dependency is `.cjs`. I didn't experience this in another repo of mine that is `"type": "commonjs"`.

Hmmm... possibly. That try/catch of `lodash`'s gets hit when there's an error from `require`.
Did you have `"type": "module"` with `0.31.0` as well? A regression would suggest that there's a race condition somewhere, as the only thing that changed was the order in which `lodash` gets imported and how frequently it gets imported.

So rpt2 does export an ESM version in its [`package.json#module`](https://github.com/ezolenko/rollup-plugin-typescript2/blob/44116044750b01c3b0d612e36f453cc8e64f79b4/package.json#L6) field etc, but that old convention was never quite supported everywhere (lots of bundlers like Webpack did support it though, but not Node itself).
You can try importing that directly instead? `import rpt2 from "rollup-plugin-typescript2/dist/rollup-plugin-typescript2.es.js"`

Related, but that's a good reminder that we should add the newer `package.json#exports` in, especially now that the spec has stabilized and _finally_ seems to be properly forward and backward-compatible (c.f. #286 where `tslib`'s `exports` broke in Node 17).

> When I finish the upgrade overall, I'll take `rollup-plugin-typescript2` off `~0.31.0` and put it back on `^0.32.0` and see if it still occurs.

Sounds good. A minimal repro would help quite a bit if you do hit it then. My suspicion is that it's something in your environment causing this, although I have no idea why it's different between `0.31.0` and `0.32.0` since it's coming from within the same version of the same bundled `lodash` code.

**Comment by shevernitskiy** [comment]
I've got the same issue after upgrading from `0.31.2` to `0.32.0`.
`Yarn 2 - v1.23.0-20220130.1630`

**Comment by agilgur5** [comment]
@shevernitskiy I see your library is open-source: https://github.com/shevernitskiy/gismeteo.
It's not a _minimal_ repro, but a repro helps a lot. I'll look into this tomorrow.

I also notice that similar to @CharlesStover , both of you were updating several dependencies at once.
And both using Yarn.

I've upgraded to rpt2 `0.32.0` in some of my other libraries and haven't encountered this as of yet, but they all use NPM fwiw.

**Comment by agilgur5** [comment]
Found your [CI log](https://github.com/shevernitskiy/gismeteo/runs/6706714333?check_suite_focus=true) as well that reproduces it.
Interestingly enough, that actually points to a file within `graphlib`, a different dependency that rpt2 uses for it's cache implementation (graph as in dependency graph) 🤔 That's also [pinned](https://github.com/ezolenko/rollup-plugin-typescript2/blob/44116044750b01c3b0d612e36f453cc8e64f79b4/package.json#L59) though and hasn't been updated in ~2 years

Hmm...

**Comment by agilgur5** [comment]
Might be due to the internal upgrade of `@rollup/plugin-commonjs` in 5a3e58b2779063b26d50c48800a80c9fd5ded726 . It's used to bundle rpt2's itself, i.e. it's own dependencies (see also #80)

**Comment by agilgur5** [comment]
Ok was able to repro the issue with both NPM and Yarn in `gismeteo`. Oddly enough though, I still can't repro it here or in my other libraries.

What I also found was that `@rollup/plugin-commonjs` was actually updated twice in this release, in the commit I mentioned above, and in 08d2f5bcad7cb9a9a0974f3707a27d29dd89da7c.

_That_ commit moves up 3 majors and in particular this issue comment stood out to me as very relevant: https://github.com/rollup/plugins/pull/1005#issuecomment-954130548 . That should be fixed by https://github.com/rollup/plugins/pull/1038 in [v22.0.0](https://github.com/rollup/plugins/blob/master/packages/commonjs/CHANGELOG.md#v2200), so I'll try updating that and seeing if it fixes this issue in `gismeteo`

**Comment by agilgur5** [comment]
Annnd yep, this exact piece of code that requires `lodash` changes in _that_ exact commit: https://github.com/ezolenko/rollup-plugin-typescript2/commit/08d2f5bcad7cb9a9a0974f3707a27d29dd89da7c#diff-80441564774ccf83c60a1616524830d5c15a9bc194b62e2c3d9f43de805c62d0L22435

Updating to `@rollup/plugin-commonjs` to v22.0.0 once again changes that piece of code and seems to fix the issue when I test with either Yarn or NPM in `gismeteo`

Can install via my [`issue-339` branch](https://github.com/agilgur5/rollup-plugin-typescript2/tree/issue-339) to workaround this issue for now. Will make a PR shortly, but have to wait on @ezolenko for a release.

In the meantime, @CharlesStover can you confirm that install my `issue-339` branch fixes this for you as well?
```sh
yarn add --dev git+https://github.com/agilgur5/rollup-plugin-typescript2#issue-339
```

**Comment by CharlesStover** [comment]
For Yarn2, the command is:

```
yarn add --dev rollup-plugin-typescript2@https://github.com/agilgur5/rollup-plugin-typescript2#head=issue-339
```

The install fails:

```
Packing rollup-plugin-typescript2@https://github.com/agilgur5/rollup-plugin-typescript2.git#commit=5f2f97d30b4a99bf6177dd3045621187666916e1 from sources
Using npm for bootstrap. Reason: found npm's "package-lock.json" lockfile

npm ERR! Unexpected token '.'
```

The NPM log for this:

```
28 verbose stack ...\nvm\v16.14.0\node_modules\npm\node_modules\libnpmfund:1
28 verbose stack ../workspaces/libnpmfund
28 verbose stack ^
28 verbose stack
28 verbose stack SyntaxError: Unexpected token '.'
28 verbose stack     at Object.compileFunction (node:vm:352:18)
28 verbose stack     at wrapSafe (node:internal/modules/cjs/loader:1032:15)
28 verbose stack     at Module._compile (node:internal/modules/cjs/loader:1067:27)
28 verbose stack     at Object.Module._extensions..js (node:internal/modules/cjs/loader:1155:10)
28 verbose stack     at Module.load (node:internal/modules/cjs/loader:981:32)
28 verbose stack     at Function.Module._load (node:internal/modules/cjs/loader:822:12)
28 verbose stack     at Module.require (node:internal/modules/cjs/loader:1005:19)
28 verbose stack     at require (node:internal/modules/cjs/helpers:102:18)
28 verbose stack     at Object.<anonymous> (...\nvm\v16.14.0\node_modules\npm\lib\utils\reify-output.js:16:38)
28 verbose stack     at Module._compile (node:internal/modules/cjs/loader:1103:14)
```

**Comment by agilgur5** [comment]
@CharlesStover that looks to be an error within your workspace given the log you posted.

rpt2 doesn't use `libnpmfund` -- I assume that is the name of the project you're working on.
```sh
❯ npm ls libnpmfund
rollup-plugin-typescript2@0.32.1 <project-dir>/rollup-plugin-typescript2
└── (empty)
```

**Comment by CharlesStover** [comment]
It is not the name of my project. 🤷‍♂️ Yarn must be using it.

**Comment by agilgur5** [comment]
> It is not the name of my project. 🤷‍♂️ Yarn must be using it.

🤷‍♂️ indeed. it might be NPM since Yarn said its "Using npm for bootstrap". Since your `npx envinfo` command failed too, it's possible your installation of NPM is corrupted -- would suggest re-installing.

In any case, that's not an error within rpt2

**Comment by mistyharsh** [comment]
I encountered similar issue. With rpt2 version `0.31.2`, the bundled js file - `<node_modules>/rollup-plugin-typescript2/dist/rollup-plugin-typescript2.cjs.js` has following:

```
var lodash;

if (typeof commonjsRequire === "function") {
  try {
    lodash = {
      clone: clone_1,
      constant: constant_1,
      each: each,
      filter: filter_1,
      has:  has_1,
      isArray: isArray_1,
      isEmpty: isEmpty_1,
      isFunction: isFunction_1,
      isUndefined: isUndefined_1,
      keys: keys_1,
      map: map_1,
      reduce: reduce_1,
      size: size_1,
      transform: transform_1,
      union: union_1,
      values: values_1
    };
  } catch (e) {
    // continue regardless of error
  }
}

if (!lodash) {
  lodash = window._;
}
```

All these functions like `reduce_1`, `transform_1`, etc. are already bundled in the file. However, with latest version `0.32.0`, the bundled file has following:

```
var lodash;

if (typeof commonjsRequire === "function") {
  try {
    lodash = {
      clone: require("lodash/clone"),
      constant: require("lodash/constant"),
      each: require("lodash/each"),
      filter: require("lodash/filter"),
      has:  require("lodash/has"),
      isArray: require("lodash/isArray"),
      isEmpty: require("lodash/isEmpty"),
      isFunction: require("lodash/isFunction"),
      isUndefined: require("lodash/isUndefined"),
      keys: require("lodash/keys"),
      map: require("lodash/map"),
      reduce: require("lodash/reduce"),
      size: require("lodash/size"),
      transform: require("lodash/transform"),
      union: require("lodash/union"),
      values: require("lodash/values")
    };
  } catch (e) {
    // continue regardless of error
  }
}

if (!lodash) {
  lodash = window._;
}
```

Since `lodash` is a dev dependency for rpt2 package, it is never installed in the user land. The node cannot handle `require("lodash/clone")` which causes it to execute `catch` clause making `lodash` variable undefined. Since it is undefined, it attempts to run `lodash = window._;` causing error at runtime.

**Comment by agilgur5** [comment]
A bugfix has been released in [`0.32.1`](https://github.com/ezolenko/rollup-plugin-typescript2/releases/0.32.1) 🎉 

I've also confirmed that `0.32.1` works inside of rpt2, in one of my own libraries, and the library that I was able to repro the bug in above, `gismeteo`.

Thanks to everyone for submitting this issue, providing repros, etc not too long after `0.32.0`'s release so it could get addressed quickly!
